File: src/core/analysis.ts
import type { Analysis, Fixation, Gaze, Recording, Settings } from "./types";
import { contains, isAutomaticAOI } from "./aoi";
import { mapTime } from "./time";
export function quantile(values: number[], q: number) {
  if (!values.length) return 0;
  const a = [...values].sort((x, y) => x - y),
    p = (a.length - 1) * q,
    i = Math.floor(p);
  return a[i] + (a[Math.min(i + 1, a.length - 1)] - a[i]) * (p - i);
}
const usable = (s: Gaze) =>
  s.valid &&
  s.x !== null &&
  s.y !== null &&
  Number.isFinite(s.x) &&
  Number.isFinite(s.y) &&
  !s.blink;
export function detectFixations(
  samples: Gaze[],
  settings: Settings,
): Fixation[] {
  const result: Fixation[] = [];
  const segments: Gaze[][] = [];
  let current: Gaze[] = [];
  for (const s of samples) {
    if (
      !usable(s) ||
      (current.length && s.t - current.at(-1)!.t > settings.maxGap)
    ) {
      if (current.length) segments.push(current);
      current = [];
    }
    if (usable(s)) current.push(s);
  }
  if (current.length) segments.push(current);
  const emit = (a: Gaze[]) => {
    if (a.length < 2 || a.at(-1)!.t - a[0].t < settings.minFixation) return;
    result.push({
      start: a[0].t,
      end: a.at(-1)!.t,
      x: a.reduce((n, s) => n + s.x!, 0) / a.length,
      y: a.reduce((n, s) => n + s.y!, 0) / a.length,
      participant: a[0].participant,
    });
  };
  for (const segment of segments) {
    if (settings.method === "ivt") {
      let run = [segment[0]];
      for (let i = 1; i < segment.length; i++) {
        const a = segment[i - 1],
          b = segment[i],
          dt = b.t - a.t,
          velocity =
            dt > 0
              ? Math.hypot(b.x! - a.x!, b.y! - a.y!) / (dt / 1e6)
              : Infinity;
        if (velocity > settings.velocity) {
          emit(run);
          run = [];
        }
        run.push(b);
      }
      emit(run);
    } else {
      let start = 0;
      while (start < segment.length - 1) {
        let end = start,
          minX = Infinity,
          maxX = -Infinity,
          minY = Infinity,
          maxY = -Infinity;
        while (end < segment.length) {
          const s = segment[end];
          const nx = Math.min(minX, s.x!),
            xx = Math.max(maxX, s.x!),
            ny = Math.min(minY, s.y!),
            xy = Math.max(maxY, s.y!);
          if (xx - nx + xy - ny > settings.dispersion) break;
          minX = nx;
          maxX = xx;
          minY = ny;
          maxY = xy;
          end++;
        }
        if (
          end > start &&
          segment[end - 1].t - segment[start].t >= settings.minFixation
        ) {
          emit(segment.slice(start, end));
          start = end;
        } else start++;
      }
    }
  }
  return result;
}
export function analyze(recording: Recording, settings: Settings): Analysis {
  const start = Math.max(0, settings.start),
    end = Math.min(recording.duration, settings.end);
  const selected = recording.samples
    .filter(
      (s) =>
        settings.participant === "all" ||
        s.participant === settings.participant,
    )
    .map((s) => ({ ...s, t: mapTime(s.t, recording.anchors) }));
  const groups = new Map<string, Gaze[]>();
  for (const s of selected) {
    const group = groups.get(s.participant) ?? [];
    group.push(s);
    groups.set(s.participant, group);
  }
  const aois = recording.aois.filter(
    (a) =>
      a.accepted && (settings.aoiScope !== "automatic" || isAutomaticAOI(a)),
  );
  const metrics = aois.map((a) => ({
    id: a.id,
    name: a.name,
    dwell: 0,
    visits: 0,
    ttff: null as number | null,
    fixationCount: 0,
    fixationDuration: 0,
  }));
  const result: Analysis = {
    samples: 0,
    validSamples: 0,
    duration: Math.max(0, end - start) * groups.size,
    validTime: 0,
    coverage: 0,
    medianHz: 0,
    gaps: 0,
    fixations: [],
    aoi: metrics,
    transitions: [],
    sequences: [],
    pupil: { mean: null, count: 0 },
    blinkSamples: 0,
    velocities: [],
    parameters: { ...settings, start, end },
    version: "gaze-studio-analysis/0.1.0",
  };
  const intervals: number[] = [],
    pupils: number[] = [],
    transitions = new Map<string, number>();
  for (const [participant, data] of groups) {
    data.sort((a, b) => a.t - b.t);
    const range = data.filter((s) => s.t >= start && s.t <= end);
    const fixes = detectFixations(range, settings);
    result.fixations.push(...fixes);
    result.samples += range.length;
    result.validSamples += range.filter(usable).length;
    result.blinkSamples += range.filter((s) => s.blink).length;
    for (const s of range)
      if (usable(s) && Number.isFinite(s.pupil) && s.pupil! > 0)
        pupils.push(s.pupil!);
    let previousHits = new Set<string>();
    let previousPrimary: string | null = null;
    for (let i = 0; i < data.length - 1; i++) {
      const a = data[i],
        b = data[i + 1];
      if (b.t <= start || a.t >= end) continue;
      const dt = b.t - a.t,
        lo = Math.max(start, a.t),
        hi = Math.min(end, b.t);
      if (dt > 0) intervals.push(dt);
      if (dt > settings.maxGap) result.gaps++;
      const valid = dt > 0 && dt <= settings.maxGap && usable(a) && usable(b);
      if (valid) {
        result.validTime += hi - lo;
        result.velocities.push({
          t: a.t,
          v: Math.hypot(b.x! - a.x!, b.y! - a.y!) / (dt / 1e6),
        });
      }
      const boundaries = [
        ...new Set([
          lo,
          hi,
          ...aois
            .flatMap((o) => [o.start, o.end])
            .filter((t) => t > lo && t < hi),
        ]),
      ].sort((a, b) => a - b);
      for (let k = 0; k < boundaries.length - 1; k++) {
        const segmentStart = boundaries[k],
          segmentEnd = boundaries[k + 1];
        const hits = new Set(
          valid
            ? aois
                .filter((o) => contains(o, a.x!, a.y!, segmentStart))
                .map((o) => o.id)
            : [],
        );
        for (const metric of metrics)
          if (hits.has(metric.id)) {
            metric.dwell += segmentEnd - segmentStart;
            if (!previousHits.has(metric.id)) metric.visits++;
          }
        const primary = hits.values().next().value ?? null;
        if (previousPrimary && primary && primary !== previousPrimary) {
          const key = JSON.stringify([previousPrimary, primary]);
          transitions.set(key, (transitions.get(key) ?? 0) + 1);
        }
        const last = result.sequences.at(-1);
        if (
          last &&
          last.participant === participant &&
          last.aoi === primary &&
          last.end === segmentStart
        )
          last.end = segmentEnd;
        else
          result.sequences.push({
            start: segmentStart,
            end: segmentEnd,
            aoi: primary,
            participant,
          });
        previousHits = hits;
        previousPrimary = primary;
      }
    }
    for (const fix of fixes)
      for (let i = 0; i < aois.length; i++)
        if (contains(aois[i], fix.x, fix.y, fix.start)) {
          const m = metrics[i];
          m.fixationCount++;
          m.fixationDuration += fix.end - fix.start;
          const ttff = fix.start - Math.max(start, aois[i].start);
          m.ttff = Math.min(m.ttff ?? Infinity, Math.max(0, ttff));
        }
  }
  result.coverage = result.duration ? result.validTime / result.duration : 0;
  result.medianHz = intervals.length ? 1e6 / quantile(intervals, 0.5) : 0;
  result.pupil = {
    mean: pupils.length
      ? pupils.reduce((a, b) => a + b, 0) / pupils.length
      : null,
    count: pupils.length,
  };
  result.transitions = [...transitions].map(([key, count]) => {
    const [from, to] = JSON.parse(key);
    return { from, to, count };
  });
  return result;
}

File: crates/gaze-core/src/analysis.rs
use crate::{
    events,
    geometry::{contains, map_time},
    types::*,
};
pub fn quantile(a: &mut [f64], q: f64) -> f64 {
    if a.is_empty() {
        return 0.;
    }
    a.sort_by(f64::total_cmp);
    let p = (a.len() - 1) as f64 * q;
    let i = p.floor() as usize;
    a[i] + (a[(i + 1).min(a.len() - 1)] - a[i]) * (p - i as f64)
}
pub fn analyze(r: Recording, mut s: Settings) -> Result<Partial, String> {
    if r.anchors.is_empty()
        || !r.duration.is_finite()
        || r.duration <= 0.
        || !["ivt", "idt"].contains(&s.method.as_str())
        || [s.velocity, s.dispersion, s.min_fixation, s.max_gap]
            .iter()
            .any(|x| !x.is_finite() || *x <= 0.)
    {
        return Err("Invalid analysis geometry, clock or settings".into());
    }
    if r.anchors
        .windows(2)
        .any(|a| a[1].gaze <= a[0].gaze || a[1].media <= a[0].media)
    {
        return Err("Clock anchors must strictly increase".into());
    }
    s.start = s.start.max(0.);
    s.end = s.end.min(r.duration);
    let mut groups: Vec<(String, Vec<Gaze>)> = Vec::new();
    let mut group_index = std::collections::HashMap::new();
    for mut g in r.samples {
        if s.participant != "all" && g.participant != s.participant {
            continue;
        }
        g.t = map_time(g.t, &r.anchors);
        let i = *group_index.entry(g.participant.clone()).or_insert_with(|| {
            groups.push((g.participant.clone(), Vec::new()));
            groups.len() - 1
        });
        groups[i].1.push(g);
    }
    let aois: Vec<Aoi> = r
        .aois
        .into_iter()
        .filter(|a| a.accepted && (s.aoi_scope.as_deref() != Some("automatic") || a.automatic()))
        .collect();
    let mut out = Analysis {
        samples: 0,
        valid_samples: 0,
        duration: (s.end - s.start).max(0.) * groups.len() as f64,
        valid_time: 0.,
        coverage: 0.,
        median_hz: 0.,
        gaps: 0,
        fixations: Vec::new(),
        aoi: aois
            .iter()
            .map(|a| Metric {
                id: a.id.clone(),
                name: a.name.clone(),
                dwell: 0.,
                visits: 0,
                ttff: None,
                fixation_count: 0,
                fixation_duration: 0.,
            })
            .collect(),
        transitions: Vec::new(),
        sequences: Vec::new(),
        pupil: Pupil {
            mean: None,
            count: 0,
        },
        blink_samples: 0,
        velocities: Vec::new(),
        parameters: s.clone(),
        version: "gaze-studio-rust/0.2.0".into(),
    };
    let mut intervals = Vec::new();
    let mut pupil_sum = 0.;
    for (participant, mut data) in groups {
        data.sort_by(|a, b| a.t.total_cmp(&b.t));
        let range: Vec<Gaze> = data
            .iter()
            .filter(|g| g.t >= s.start && g.t <= s.end)
            .cloned()
            .collect();
        let fixes = events::detect(&range, &s);
        out.samples += range.len();
        out.valid_samples += range.iter().filter(|g| g.usable()).count();
        out.blink_samples += range.iter().filter(|g| g.blink).count();
        for g in &range {
            if g.usable() {
                if let Some(p) = g.pupil {
                    if p.is_finite() && p > 0. {
                        pupil_sum += p;
                        out.pupil.count += 1;
                    }
                }
            }
        }
        let mut previous_hits = vec![false; aois.len()];
        let mut previous_primary: Option<String> = None;
        for pair in data.windows(2) {
            let a = &pair[0];
            let b = &pair[1];
            if b.t <= s.start || a.t >= s.end {
                continue;
            }
            let dt = b.t - a.t;
            let lo = a.t.max(s.start);
            let hi = b.t.min(s.end);
            if dt > 0. {
                intervals.push(dt)
            }
            if dt > s.max_gap {
                out.gaps += 1
            }
            let valid = dt > 0. && dt <= s.max_gap && a.usable() && b.usable();
            if valid {
                out.valid_time += hi - lo;
                out.velocities.push(Velocity {
                    t: a.t,
                    v: (b.x.unwrap() - a.x.unwrap()).hypot(b.y.unwrap() - a.y.unwrap())
                        / (dt / 1e6),
                });
            }
            let mut boundaries = vec![lo, hi];
            for o in &aois {
                for t in [o.start, o.end] {
                    if t > lo && t < hi {
                        boundaries.push(t)
                    }
                }
            }
            boundaries.sort_by(f64::total_cmp);
            boundaries.dedup();
            for span in boundaries.windows(2) {
                let hits: Vec<bool> = aois
                    .iter()
                    .map(|o| valid && contains(o, a.x.unwrap(), a.y.unwrap(), span[0]))
                    .collect();
                for (i, hit) in hits.iter().enumerate() {
                    if *hit {
                        out.aoi[i].dwell += span[1] - span[0];
                        if !previous_hits[i] {
                            out.aoi[i].visits += 1
                        }
                    }
                }
                let primary = hits.iter().position(|v| *v).map(|i| aois[i].id.clone());
                if let (Some(from), Some(to)) = (&previous_primary, &primary) {
                    if from != to {
                        if let Some(old) = out
                            .transitions
                            .iter_mut()
                            .find(|v| &v.from == from && &v.to == to)
                        {
                            old.count += 1
                        } else {
                            out.transitions.push(Transition {
                                from: from.clone(),
                                to: to.clone(),
                                count: 1,
                            });
                        }
                    }
                }
                if let Some(last) = out.sequences.last_mut().filter(|v| {
                    v.participant == participant && v.aoi == primary && v.end == span[0]
                }) {
                    last.end = span[1]
                } else {
                    out.sequences.push(Sequence {
                        start: span[0],
                        end: span[1],
                        aoi: primary.clone(),
                        participant: participant.clone(),
                    })
                }
                previous_hits = hits;
                previous_primary = primary;
            }
        }
        for f in &fixes {
            for (i, a) in aois.iter().enumerate() {
                if contains(a, f.x, f.y, f.start) {
                    let m = &mut out.aoi[i];
                    m.fixation_count += 1;
                    m.fixation_duration += f.end - f.start;
                    let ttff = (f.start - s.start.max(a.start)).max(0.);
                    m.ttff = Some(m.ttff.unwrap_or(f64::INFINITY).min(ttff));
                }
            }
        }
        out.fixations.extend(fixes);
    }
    out.coverage = if out.duration > 0. {
        out.valid_time / out.duration
    } else {
        0.
    };
    out.median_hz = if intervals.is_empty() {
        0.
    } else {
        1e6 / quantile(&mut intervals.clone(), 0.5)
    };
    out.pupil.mean = if out.pupil.count > 0 {
        Some(pupil_sum / out.pupil.count as f64)
    } else {
        None
    };
    Ok(Partial {
        result: out,
        intervals,
        pupil_sum,
    })
}
pub fn merge(mut parts: Vec<Partial>) -> Result<Analysis, String> {
    if parts.is_empty() {
        return Err("No analysis partitions".into());
    }
    let mut first = parts.remove(0);
    for p in parts {
        let a = &mut first.result;
        let b = p.result;
        if a.aoi.iter().map(|m| &m.id).ne(b.aoi.iter().map(|m| &m.id)) {
            return Err("Incompatible AOI partitions".into());
        }
        a.samples += b.samples;
        a.valid_samples += b.valid_samples;
        a.duration += b.duration;
        a.valid_time += b.valid_time;
        a.gaps += b.gaps;
        a.blink_samples += b.blink_samples;
        a.fixations.extend(b.fixations);
        a.sequences.extend(b.sequences);
        a.velocities.extend(b.velocities);
        a.pupil.count += b.pupil.count;
        for (m, n) in a.aoi.iter_mut().zip(b.aoi) {
            m.dwell += n.dwell;
            m.visits += n.visits;
            m.fixation_count += n.fixation_count;
            m.fixation_duration += n.fixation_duration;
            if let Some(t) = n.ttff {
                m.ttff = Some(m.ttff.unwrap_or(f64::INFINITY).min(t));
            }
        }
        for t in b.transitions {
            if let Some(old) = a
                .transitions
                .iter_mut()
                .find(|v| v.from == t.from && v.to == t.to)
            {
                old.count += t.count
            } else {
                a.transitions.push(t)
            }
        }
        first.intervals.extend(p.intervals);
        first.pupil_sum += p.pupil_sum;
    }
    let a = &mut first.result;
    a.coverage = if a.duration > 0. {
        a.valid_time / a.duration
    } else {
        0.
    };
    a.median_hz = if first.intervals.is_empty() {
        0.
    } else {
        1e6 / quantile(&mut first.intervals, 0.5)
    };
    a.pupil.mean = if a.pupil.count > 0 {
        Some(first.pupil_sum / a.pupil.count as f64)
    } else {
        None
    };
    Ok(first.result)
}
